lms/djangoapps/django_comment_client/utils.py

```diff
@@ -602,53 +606,51 @@ def get_user_group_ids(course_id, content, user=None):
     return user_group_id, content_user_group_id
 
 
-def get_annotated_content_info(course_id, content, user, user_info):
-    """
-    Get metadata for an individual content (thread or comment)
-    """
-    voted = ''
-    if content['id'] in user_info['upvoted_ids']:
-        voted = 'up'
-    elif content['id'] in user_info['downvoted_ids']:
-        voted = 'down'
-    return {
-        'voted': voted,
-        'subscribed': content['id'] in user_info['subscribed_thread_ids'],
-        'ability': get_ability(course_id, content, user),
-    }
-
-# TODO: RENAME
-
-
-def get_annotated_content_infos(course_id, thread, user, user_info):
-    """
-    Get metadata for a thread and its children
-    """
-    infos = {}
-
-    def annotate(content):
-        infos[str(content['id'])] = get_annotated_content_info(course_id, content, user, user_info)
-        for child in (
-                content.get('children', []) +
-                content.get('endorsed_responses', []) +
-                content.get('non_endorsed_responses', [])
-        ):
-            annotate(child)
-    annotate(thread)
-    return infos
-
-
 def get_metadata_for_threads(course_id, threads, user, user_info):
     """
     Returns annotated content information for the specified course, threads, and user information
     """
+    _thread_usernames = set()
+    course_discussion_settings = get_course_discussion_settings(course_id)
 
-    def infogetter(thread):
-        return get_annotated_content_infos(course_id, thread, user, user_info)
-
-    metadata = {}
+    def _flatten_content_usernames(thread):
+        """
+        recursively turn a tree into a set to make metadata extraction easier
+        """
+        _thread_usernames.add(thread['username'])
+        children = thread.get('children', []).extend(
+            thread.get('endorsed_responses', [])
+        ).extend(
+            thread.get('non_endorsed_responses', [])
+        )
+        for child in children:
+            _flatten_content_usernames(child)
     for thread in threads:
-        metadata.update(infogetter(thread))
+        _flatten_content_usernames(thread)
+
+    content_user_group_ids = {
+        name: get_group_id_for_user(
+            get_user_by_username_or_email(name),
+            course_discussion_settings
+        ) for name in _thread_usernames
+    }
+    user_group_id = get_group_id_for_user(user, course_discussion_settings)
```

Comment:
You might be able to avoid this call if the user is already included in `content_user_group_ids`.

Reply:
(with the caveat that I'm only 80% clear on what all is going on, here)